### PR TITLE
External services validation with JSON Schema

### LINKED ISSUE #1721: Require valid external service config json before creating/updating external services.
Right now, external service config json is validated on the frontend (which yields green squiggles in the frontend). The existence of validation errors doesn't prevent the form from being submitted however; ideally it would.

Either we need to do validation on the backend (e.g. https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/pkg/conf/validate.go#L76:11), or, if possible and easier, piggy back on the validation already being done in the client by the monaco editor and prevent form submission there (e.g. does this help? https://github.com/Microsoft/vscode/pull/31045).

cc @tsenart since this relates to external services and @slimsag since this relates to backend validation which is already being done for site config.

### PR DESCRIPTION
**Pull request by tsenart**
Fixes #1721

### CODE REVIEW

File: cmd/frontend/db/external_services.go [hunk 1]
```diff
@@ -45,100 +66,118 @@ func (o ExternalServicesListOptions) sqlConditions() []*sqlf.Query {
...
+		fmt.Sprintf(`{"$ref": "site.schema.json#/definitions/%s"}`, ext.Definition),
+	))
+
+	if err != nil {
+		return errors.Wrapf(err, "failed to compile schema for external service of kind %q", kind)
+	}
+
+	normalized, err := jsonc.Parse(config)
+	if err != nil {
+		return errors.Wrapf(err, "failed to normalize JSON")
```
Comment (nicksnyder) [inline cmd/frontend/db/external_services.go]:
```suggestion
		return errors.Wrap(err, "failed to normalize JSON")
```

**Comment by tsenart** [comment]
This is not ready for review yet as the title of the PR indicates @sqs, @nicksnyder.

**Comment by tsenart** [comment]
@nicksnyder, @keegancsmith: This is now ready for review. Please take a look.

**Comment by tsenart** [comment]
Also @beyang, please take a look since I touched some authz code.

File: schema/site.schema.json [hunk 2]
```diff
@@ -535,13 +566,16 @@
       "description": "Connection to Git repositories for which an external service integration isn't yet available.",
       "type": "object",
       "required": ["repos"],
-      "additionalProperties": false,
```
Comment (nicksnyder) [inline schema/site.schema.json]: why was this removed?

Comment (tsenart) [reply schema/site.schema.json]: That's a left-over from some experiments. Thanks for catching!

File: cmd/frontend/db/external_services.go [hunk 3]
```diff
@@ -4,27 +4,47 @@ import (
...
+	"GITLAB":          {CodeHost: true, Definition: "GitLabConnection"},
+	"GITOLITE":        {CodeHost: true, Definition: "GitoliteConnection"},
+	"PHABRICATOR":     {CodeHost: true, Definition: "PhabricatorConnection"},
+	"OTHER":           {CodeHost: true, Definition: "OtherExternalServiceConnection"},
+}
+
+// ExternalServiceKind describes a kind of external service.
+type ExternalServiceKind struct {
+	// True if the external service can host repositories.
+	CodeHost bool
```
Comment (keegancsmith) [inline cmd/frontend/db/external_services.go]: Out of interest what is an example of CodeHost that is false. I see we had this field in our older code as well.

File: cmd/frontend/db/external_services.go [hunk 4]
```diff
@@ -45,106 +65,126 @@ func (o ExternalServicesListOptions) sqlConditions() []*sqlf.Query {
...
-	case "BITBUCKETSERVER":
-		var cfg schema.BitbucketServerConnection
-		if err := jsonc.Unmarshal(config, &cfg); err != nil {
-			return err
-		}
-		if cfg.Token != "" && cfg.Password != "" {
-			return errors.New("Specify either a token or a username/password, but not both")
-		} else if cfg.Token == "" && cfg.Username == "" && cfg.Password == "" {
-			return errors.New("Specify either a token or a username/password to authenticate")
-		}
```
Comment (keegancsmith) [inline cmd/frontend/db/external_services.go]: so our schema lib does the above validation now?

File: cmd/frontend/db/external_services.go (hunk 3 above)
Comment (tsenart) [reply cmd/frontend/db/external_services.go]: Not yet. But the rationale for the introduction of the "External Service" concept is in part that we'll one day support other kinds of external services. To be seen if this was premature abstraction or not.

File: cmd/frontend/db/external_services.go (hunk 4 above)
Comment (tsenart) [reply cmd/frontend/db/external_services.go]: Yes. This is how we do it: https://github.com/sourcegraph/sourcegraph/pull/2017/files#diff-46a0ae99a143fd2b76b25309eba01a47R387

And here are the test cases for it: https://github.com/sourcegraph/sourcegraph/pull/2017/files#diff-162e709517e3585ef8bd0dd0825bbec8R152

**Comment by tsenart** [comment]
@beyang: I'm confident that the change I've done to the authz code is safe so I'm merging this now. Please review this when you have availability. We can iterate if you happen to have concerns / feedback.
